src/python/pants/backend/java/compile/javac.py

```diff
@@ -68,6 +75,21 @@ async def compile_java_source(
             exit_code=1,
         )
 
+    # Capture just the `ClasspathEntry` objects that are listed as `export` types by source analysis
+    deps_to_classpath_entries = dict(
+        zip(request.component.dependencies, direct_dependency_classpath_entries or ())
+    )
+    # Re-request inferred dependencies to get a list of export dependency addresses
+    inferred_dependencies = await MultiGet(
+        Get(JavaInferredDependencies, JavaInferredDependenciesAndExportsRequest(tgt.address))
+        for tgt in request.component.members
+        if JavaFieldSet.is_applicable(tgt)
+    )
+    flat_exports = (export for i in inferred_dependencies for export in i.exports)
+    export_targets = await Get(CoarsenedTargets, Addresses(flat_exports))
```

Comment:
Recalculating the `CoarsenedTargets` isn't cheap, since the computation has to walk into transitive deps.

Instead, it seems like this should be able to filter `deps_to_classpath_entries` to just classpath entries containing exported targets with something like:
```python
export_classpath_entries = [
  cpe
  for coarsened_target, cpe in deps_to_classpath_entries.items()
  if any(m.address in flat_exports for m in coarsened_target.members)
]
```

Reply:
Nice! This appears to be equivalent in Guice, which is Good Enough™ for me.